File: src/cwlint/extractors/kernel_env_allowlist.py
"""Kernel-env allowlist (`MGn` in v1.6259.1) — strings the VM kernel-shell sees.

Anchor: `new Set([...])` literal containing the unique combination "HOME",
"PATH", and "OPERON_SECRET_VARS" — that triple uniquely identifies MGn.
"""

from __future__ import annotations

import re
from typing import Any

from cwlint.extractors._base import Extractor

_ANCHOR = re.compile(
    r"""new\s+Set\(\[
        (?P<body>[^\]]*"OPERON_SECRET_VARS"[^\]]*)
    \]\)""",
    re.VERBOSE | re.DOTALL,
)
_STRING = re.compile(r'"([A-Za-z_][A-Za-z0-9_]*)"')


class KernelEnvAllowlistExtractor(Extractor):
    field_name = "kernel_env_passthrough"
    target_bundle = "desktop"

    def extract(self, source: str) -> dict[str, Any] | None:
        for m in _ANCHOR.finditer(source):
            body = m["body"]
            if "HOME" in body and "PATH" in body:
                names = _STRING.findall(body)
                if names and "OPERON_SECRET_VARS" in names:
                    return {
                        "allowlist": names,
                        "deleted_after_filter": [
                            n for n in ("HOME", "USER", "LOGNAME", "TMPDIR") if n in names
                        ],
                    }
        return None

File: src/extractors/kernel-env-allowlist.ts
/**
 * Kernel-env allowlist extractor (`MGn`-style symbol in v1.6259+).
 *
 * Anchor: a `new Set([...])` literal whose body contains the unique triple
 * "HOME", "PATH", "OPERON_SECRET_VARS". That triple uniquely identifies
 * `MGn` across observed Claude.app builds.
 *
 * Output shape (matches Python reference):
 *   {
 *     allowlist: string[],
 *     deleted_after_filter: string[]   // intersection with HOME/USER/LOGNAME/TMPDIR
 *   }
 */

import type { NodePath } from "@babel/traverse";
import * as t from "@babel/types";

import type { ExtractContext } from "./_ast.js";
import { traverse } from "./_ast.js";

const DELETED_AFTER_FILTER_PROBE = ["HOME", "USER", "LOGNAME", "TMPDIR"] as const;

export interface KernelEnvAllowlistFragment {
  allowlist: string[];
  deleted_after_filter: string[];
}

export function extractKernelEnvAllowlist(ctx: ExtractContext): KernelEnvAllowlistFragment | null {
  let result: KernelEnvAllowlistFragment | null = null;

  traverse(ctx.ast, {
    NewExpression(path: NodePath<t.NewExpression>) {
      if (result !== null) {
        path.stop();
        return;
      }
      const callee = path.node.callee;
      if (!t.isIdentifier(callee) || callee.name !== "Set") return;
      const arg = path.node.arguments[0];
      if (arg === undefined || !t.isArrayExpression(arg)) return;

      // Cheap pre-filter: the literal-string members must include the unique
      // triple. We walk the elements once and collect the literal names.
      const names: string[] = [];
      for (const el of arg.elements) {
        if (el === null) continue;
        if (t.isStringLiteral(el)) names.push(el.value);
      }
      if (
        !names.includes("HOME") ||
        !names.includes("PATH") ||
        !names.includes("OPERON_SECRET_VARS")
      ) {
        return;
      }

      result = {
        allowlist: names,
        deleted_after_filter: DELETED_AFTER_FILTER_PROBE.filter((n) => names.includes(n)),
      };
      path.stop();
    },
  });

  return result;
}

extractKernelEnvAllowlist.fieldName = "kernel_env_passthrough" as const;
extractKernelEnvAllowlist.targetBundle = "desktop" as const;
